pkg/cloudprovider/vsphere/nodemanager_test.go

```diff
@@ -184,6 +185,97 @@ func TestAlphaDualStack(t *testing.T) {
 
 	ips := nodeList[0].Addresses
 
+	ipv4Ips := returnIPsFromSpecificFamily(vcfg.IPv4Family, ips)
+	size := len(ipv4Ips)
+	// size need to equal to 2, one for internal and one for external, in this case they are the same
+	if size != 2 {
+		t.Errorf("Should only return single IPv4 address. expected: 1, actual: %d", size)
+	} else if !strings.EqualFold(ipv4Ips[0], ipv4Ip) {
+		t.Errorf("IPv6 does not match. expected: %s, actual: %s", ipv4Ip, ipv4Ips[0])
+	}
+
+	ipv6Ips := returnIPsFromSpecificFamily(vcfg.IPv6Family, ips)
+	size = len(ipv6Ips)
+	if size != 2 {
+		t.Errorf("Should only return single IPv6 address. expected: 1, actual: %d", size)
+	} else if !strings.EqualFold(ipv6Ips[0], ipv6Ip) {
+		t.Errorf("IPv6 does not match. expected: fd01:0:101:2609:bdd2:ee20:7bd7:5836, actual: %s", ipv6Ips[0])
+	}
+}
+
+func TestExternalInternalSubnet(t *testing.T) {
+
+	var subnetCidrINIConfig = `
+	[VirtualCenter "0.0.0.0"]
+	user = user
+	password = password
+	ip-family = ipv6,ipv4
+	[Nodes]
+	internal-network-subnet-cidr = "10.0.0.0/24"
+	external-network-subnet-cidr = "fd01:0:101:2609:bdd2:ee20:7bd7:0/112"
+	`
+
+	iniCfg, err := ccfg.ReadCPIConfigINI([]byte(subnetCidrINIConfig))
+	if err != nil{
+		t.Errorf("Seeing error %s when import INI config", err)
+	}
+	cfg, ok := configFromEnvOrSim(true)
+	defer ok()
+	iniCfg.Config = *cfg
+
+	connMgr := cm.NewConnectionManager(cfg, nil, nil)
+	defer connMgr.Logout()
+
+	ipv4Ip := "10.0.0.1"
+	ipv6Ip := "fd01:0:101:2609:bdd2:ee20:7bd7:5836"
+
+	nm := newNodeManager(iniCfg, connMgr)
+
+	vm := simulator.Map.Any("VirtualMachine").(*simulator.VirtualMachine)
+	vm.Guest.HostName = strings.ToLower(vm.Name) // simulator.SearchIndex.FindByDnsName matches against the guest.hostName property
+	vm.Guest.Net = []vimtypes.GuestNicInfo{
+		{
+			Network:   "foo-bar",
+			IpAddress: []string{ipv4Ip, ipv6Ip},
+		},
+	}
+
+	err = connMgr.Connect(context.Background(), connMgr.VsphereInstanceMap[cfg.Global.VCenterIP])
+	if err != nil {
+		t.Errorf("Failed to Connect to vSphere: %s", err)
+	}
+
+	// set config for ip for vc to ipv4, ipv6 (dual-stack)
+	vcInstance := nm.connectionManager.VsphereInstanceMap[cfg.Global.VCenterIP]
+	vcInstance.Cfg.IPFamilyPriority = []string{"ipv6", "ipv4"}
+
+	name := vm.Name
+	UUID := vm.Config.Uuid
+	k8sUUID := ConvertK8sUUIDtoNormal(UUID)
+
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Status: v1.NodeStatus{
+			NodeInfo: v1.NodeSystemInfo{
+				SystemUUID: k8sUUID,
+			},
+		},
+	}
+
+	// get node registered so node can be exported
+	nm.RegisterNode(node)
+	err = nm.DiscoverNode(name, cm.FindVMByName)
+	if err != nil {
+		t.Errorf("Failed DiscoverNode: %s", err)
+	}
+
+	nodeList := make([]*pb.Node, 0)
+	_ = nm.ExportNodes("", "", &nodeList)
```

Comment:
Updating datacenterToNodeList makes this test pass since we're calling `ExportNodes`.  But does the node registration code path call this? IIRC this is only called by the internal gRPC server which is not on by default.

Reply:
IIRC the gRPC server was an experimental feature that was never turned on by default.

Reply:
My understanding of the purpose of calling `ExportNodes` is to get the Nodelist so I can check against the test case. I don't think that the node registration code path calls this function. I only found one use case for `ExportNodes` which is [here](https://github.com/kubernetes/cloud-provider-vsphere/blob/3058adc81bd0a3aa59517b3fb2160f04a81bb0ac/pkg/cloudprovider/vsphere/server/server.go#L87).

Reply:
@andrewsykim That is correct. The code is there to support a gRPC endpoint which is experimental and is disabled by default.